refactor(localization_error_monitor): refactor diag

YamatoAndo opened the pull request:
## Description

I refactored localization_error_monitor's diagnostics.

- refactor localization_error_monitor's diagnostics code
- add test code for diagnostics code
- add the diagnostics image  to the readme
- rename diagnostic_aggregator's param
 
And I have consolidated what was originally output separately into one.

**before**

![Screenshot from 2023-09-12 18-46-47](https://github.com/autowarefoundation/autoware.universe/assets/13819566/7a3db351-297d-4277-999c-6a38ec766eaf)

![Screenshot from 2023-09-12 18-46-56](https://github.com/autowarefoundation/autoware.universe/assets/13819566/cbde9f57-16b4-4f78-bf51-32a2c1891365)


-----

**after**
![diagnostics](https://github.com/autowarefoundation/autoware.universe/assets/13819566/0f835a4a-d4dc-4ff7-805f-d13443375b81)


## Related links

N/A
<!-- Write the links related to this PR. Private links should be clearly marked as private, for example, '[FOO COMPANY INTERNAL LINK](https://example.com)'. -->

## Tests performed

- LSim works well.
- check localization_error_monitor's diagnostics by using `ros2 run rqt_runtime_monitor rqt_runtime_monitor`



<!-- Describe how you have tested this PR. -->

## Notes for reviewers

N/A

<!-- Write additional information if necessary. It should be written if there are related PRs that should be merged at the same time. -->

## Interface changes

N/A

<!-- Describe any changed interfaces, such as topics, services, or parameters. -->

## Effects on system behavior

N/A
<!-- Describe how this PR affects the system behavior. -->

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

KYabuuchi commented on localization/localization_error_monitor/src/diagnostics.cpp:
```diff
@@ -0,0 +1,91 @@
+  diagnostic_msgs::msg::DiagnosticStatus merged_stat;
+
+  for (const auto & stat : stat_array) {
+    if ((stat.level > 0) && (merged_stat.level > 0)) {
+      if (!merged_stat.message.empty()) {
+        merged_stat.message += "; ";
+        merged_stat.message += stat.message;
+      }
+    } else if (stat.level > merged_stat.level) {
+      merged_stat.message = stat.message;
+    }
```
In the original implementation, the condition that goes into line 80 is equivalent to `stat.level>0`, so it can be rewritten as follows.  Also, replacing 0 with Status::OK is easier to understand.
```suggestion
    if ((stat.level > diagnostic_msgs::msg::DiagnosticStatus::OK)) {
      if (!merged_stat.message.empty()) {
        merged_stat.message += "; ";
      }
      merged_stat.message += stat.message;
    }
```

YamatoAndo replied:
You're a genius!

KYabuuchi commented on localization/localization_error_monitor/src/diagnostics.cpp:
```diff
@@ -0,0 +1,91 @@
+  return stat;
+}
+
+diagnostic_msgs::msg::DiagnosticStatus mergeDiagnosticStatus(
+  const std::vector<diagnostic_msgs::msg::DiagnosticStatus> & stat_array)
```
Could you add comments to explain how the merging process will work.
For example,  "The highest level within the stat_array will be reflected in the merged_stat. When all stat_array entries are 'OK,' the message of merged_stat will be empty."

YamatoAndo replied:
fixed https://github.com/autowarefoundation/autoware.universe/pull/4964/commits/16e197d5c188f498e4a567e4a70ad610238ac199

KYabuuchi reviewed (commented):
This refactoring is basically good. But some parts are difficult to read.
Please check my following comments.

KYabuuchi commented:
We need to modify [`system/system_error_monitor/config/diagnostic_aggregator/localization.param.yaml`](https://github.com/autowarefoundation/autoware.universe/blob/535537db1ab593aaff931c0866f4c92d05e21695/system/system_error_monitor/config/diagnostic_aggregator/localization.param.yaml#L27-L31) too.

The condition needs to be changed from `contains: [": localization_accuracy"]` to `contains: ["localization_accuracy"]` in order to aggregate the concatenated diagnostics.

YamatoAndo commented:
@ito-san 
I changed the component name from `localization_error_monitor: localization_accuracy` to `localization: localization_error_monitor`, so I updated the diagnostic_aggregator's param. 
Please review this PR.

YamatoAndo commented:
@KYabuuchi I have made a change so that when the diagnostic status is "OK," the message "OK" will be displayed.

SakodaShintaro reviewed (approved):
Looks Good To Me

KYabuuchi reviewed (approved):
LGTM

KYabuuchi commented:
RELATED LINK: https://github.com/autowarefoundation/autoware_launch/pull/565 
(It update default parameter of localization_error_monitor)

ito-san reviewed (approved):
LGTM

YamatoAndo commented:
@Motsu-san Could you please approve this pr?

Motsu-san commented on localization/localization_error_monitor/config/localization_error_monitor.param.yaml:
```diff
@@ -4,4 +4,4 @@
     warn_ellipse_size: 0.8
     error_ellipse_size_lateral_direction: 0.3
-    warn_ellipse_size_lateral_direction: 0.2
+    warn_ellipse_size_lateral_direction: 0.25
```
This seems to be a **fix**. Why was this changed?

YamatoAndo replied:
To be the same value as [this](https://github.com/autowarefoundation/autoware_launch/blob/main/autoware_launch/config/localization/localization_error_monitor.param.yaml#L7).

Motsu-san replied:
Do you mean 0.25 is right? Or is this match with autoware_launch important and can it be changed whether 0.25 or whatever as users like ?

Motsu-san replied:
This is the latter. I got the reply directly.

Motsu-san commented on localization/localization_error_monitor/src/node.cpp:
```diff
@@ -40,7 +41,7 @@ LocalizationErrorMonitor::LocalizationErrorMonitor()
     this->declare_parameter("error_ellipse_size_lateral_direction", 0.3);
   warn_ellipse_size_lateral_direction_ =
-    this->declare_parameter("warn_ellipse_size_lateral_direction", 0.2);
+    this->declare_parameter("warn_ellipse_size_lateral_direction", 0.25);
```
Same as [this](https://github.com/autowarefoundation/autoware.universe/pull/4964/files#diff-9ea4b7dd951014d6499494a7e340fca96ec106bc747f525e655f6acbe94d8954R7).

Motsu-san reviewed (commented):
I would like to understand why this change is needed and how influences on localization_error_monitor.

Motsu-san commented:
@YamatoAndo After you resolve my question and conflicts, I will approve.

Motsu-san reviewed (approved):
[![LGTMeow](https://lgtm-images.lgtmeow.com/2022/03/20/00/401b99bb-179c-41ca-a44f-979b3e2e1490.webp)](https://lgtmeow.com)
